src/Google.Pubsub.V1/SimpleSubscription.cs

```diff
@@ -0,0 +1,178 @@
+
+        /// <summary>
+        /// The parent object of this subscription.
+        /// </summary>
+        public SimplePubsub Parent { get; }
+
+        /// <summary>
+        /// Constructs an instance with the given parent, for the given subscription ID which is expected to
+        /// subscribed to the given topic ID in the same parent project.
+        /// </summary>
+        /// <remarks>
+        /// This constructor does not validate that the subscription exists, or that it is actually subscribed to
+        /// the given topic.
+        /// </remarks>
+        /// <param name="parent">The parent object, usually used to create the new instance. Must not be null.</param>
+        /// <param name="subscriptionId">The ID of the subscription within the parent project. Must not be null.</param>
+        /// <param name="topicId">The ID of the topic within the parent project. Must not be null.</param>
+        public SimpleSubscription(SimplePubsub parent, string subscriptionId, string topicId)
+        {
+            Parent = GaxPreconditions.CheckNotNull(parent, nameof(parent));
+            SubscriptionId = GaxPreconditions.CheckNotNull(subscriptionId, nameof(subscriptionId));
+            TopicId = GaxPreconditions.CheckNotNull(topicId, nameof(topicId));
+            ResourceName = GetSubscriptionName(parent.ProjectId, subscriptionId);
+            TopicResourceName = GetTopicName(parent.ProjectId, topicId);
+        }
+
+        // TODO:
+        // - Do we actually want Poll? Unclear how much value it adds.
+        // - Should PullAndAcknowledge take an action, and only acknowledge messages after acting on them?
+        //   - If so, should it acknowledge in batches, or individually? If it does it in batches, should it
+        //     acknowledge successful messages in a finally block if a later message failed?
+
+        /// <summary>
+        /// Polls the subscription until the specified cancellation token is cancelled, acknowledging
+        /// messages automatically and executing the given callback for each one.
+        /// </summary>
+        /// <remarks>
+        /// <para>
+        /// When the cancellation token is cancelled, this method returns without throwing an exception. It is
+        /// possible that messages will already have been acknowledged but not processed when that occurs.
+        /// </para>
+        /// </remarks>
+        /// <param name="action">The callback to execute for each message. Must not be null.</param>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests.</param>
+        public void Poll(Action<PubsubMessage> action, CancellationToken cancellationToken)
+        {
+            while (true)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    return;
+                }
+                try
+                {
+                    var messages = PullAndAcknowledge(cancellationToken);
```

Comment:
Bit worried about this pattern, since you are acknowledging the message right away you are negating the effect of retries. Since `action` is called _after_ the acknowledgment of the message if it fails in any way the message will not be re-delivered. Looking at [this page](https://cloud.google.com/pubsub/subscriber) it seems that the acknowledgement should only happen _after_ the processing of the message is done.